File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Tuvalu.m
//
//  UIImage+Tuvalu.m
//  CountriesFlags
//

#import "Countries/UIImage+Tuvalu.h"

#import "Countries/UIImage+UnitedKingdom.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"
#import "Extensions/CGSize+Extension.h"

@implementation UIImage (Tuvalu)

+ (UIImage * _Nullable)flagTuvaluWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:156.0f/255.0f
                                          blue:222.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:254.0f/255.0f
                                           green:221.0f/255.0f
                                            blue:0.0f
                                           alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    // Nanumea island
    CGFloat starRadius = size.height / 100.0f * 8.7f;
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 1000.0f * 546.0f, size.height / 100.0f * 90.0f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // Nukufetau island
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 1000.0f * 842.0f, size.height / 100.0f * 64.15f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // Niulakita island
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 1000.0f * 929.0f, size.height / 100.0f * 53.25f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // Nukulaelae island
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 1000.0f * 929.0f, size.height / 100.0f * 13.25f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // Niutao island
    CGPoint starCenter = CGPointMake(size.width / 1000.0f * 650.0f, size.height / 100.0f * 64.25f);
    UIBezierPath* path = [UIBezierPath starAtPoint:starCenter radius:starRadius type:FlagStarTypeFivePointed];
    [path rotateToAngle:36.0f center:starCenter];
    [path fill];
    // Nanumanga island
    starCenter = CGPointMake(size.width / 1000.0f * 650.0f, size.height / 100.0f * 85.0f);
    path = [UIBezierPath starAtPoint:starCenter
                              radius:starRadius
                                type:FlagStarTypeFivePointed];
    [path rotateToAngle:36.0f center:starCenter];
    [path fill];
    // Nui island
    starCenter = CGPointMake(size.width / 1000.0f * 754.5f, size.height / 100.0f * 79.25f);
    path = [UIBezierPath starAtPoint:starCenter
                              radius:starRadius
                                type:FlagStarTypeFivePointed];
    [path rotateToAngle:36.0f center:starCenter];
    [path fill];
    // Vaitupu island
    starCenter = CGPointMake(size.width / 1000.0f * 754.5f, size.height / 100.0f * 35.75f);
    path = [UIBezierPath starAtPoint:starCenter
                              radius:starRadius
                                type:FlagStarTypeFivePointed];
    [path rotateToAngle:36.0f center:starCenter];
    [path fill];
    // Funafuti island
    starCenter = CGPointMake(size.width / 1000.0f * 841.5f, size.height / 100.0f * 29.3f);
    path = [UIBezierPath starAtPoint:starCenter
                              radius:starRadius
                                type:FlagStarTypeFivePointed];
    [path rotateToAngle:36.0f center:starCenter];
    [path fill];

    [[UIImage flagUnitedKingdomWithSize:CGSizeHalfSize(size)] drawAtPoint:CGPointZero];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Tuvalu.swift
//
//  UIImage+Tuvalu.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Tuvalu "wikipedia"
     
     Generates the flag of Tuvalu.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Tuvalu.
     */
    internal static func flagTuvalu(with size: CGSize) -> UIImage {
        let colorBlue = UIColor(red: 0.0,
                                green: 156.0 / 255.0,
                                blue: 222.0 / 255.0,
                                alpha: 1.0)
        let colorYellow = UIColor(red: 254.0 / 255.0,
                                  green: 221.0 / 255.0,
                                  blue: 0.0,
                                  alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        // Nanumea island
        let starRadius = size.height / 100.0 * 8.7
        UIBezierPath.star(at: CGPoint(x: size.width / 1000.0 * 546.0, y: size.height / 100.0 * 90.0),
                          radius: starRadius).fill()
        // Nukufetau island
        UIBezierPath.star(at: CGPoint(x: size.width / 1000.0 * 842.0, y: size.height / 100.0 * 64.15),
                          radius: starRadius).fill()
        // Niulakita island
        UIBezierPath.star(at: CGPoint(x: size.width / 1000.0 * 929.0, y: size.height / 100.0 * 53.25),
                          radius: starRadius).fill()
        // Nukulaelae island
        UIBezierPath.star(at: CGPoint(x: size.width / 1000.0 * 929.0, y: size.height / 100.0 * 13.25),
                          radius: starRadius).fill()
        // Niutao island
        var starCenter = CGPoint(x: size.width / 1000.0 * 650.0,
                                 y: size.height / 100.0 * 64.25)
        var path = UIBezierPath.star(at: starCenter, radius: starRadius)
        path.rotate(to: 36.0, around: starCenter)
        path.fill()
        // Nanumanga island
        starCenter = CGPoint(x: size.width / 1000.0 * 650.0, y: size.height / 100.0 * 85.0)
        path = UIBezierPath.star(at: starCenter, radius: starRadius)
        path.rotate(to: 36.0, around: starCenter)
        path.fill()
        // Nui island
        starCenter = CGPoint(x: size.width / 1000.0 * 754.5, y: size.height / 100.0 * 79.25)
        path = UIBezierPath.star(at: starCenter, radius: starRadius)
        path.rotate(to: 36.0, around: starCenter)
        path.fill()
        // Vaitupu island
        starCenter = CGPoint(x: size.width / 1000.0 * 754.5, y: size.height / 100.0 * 35.75)
        path = UIBezierPath.star(at: starCenter, radius: starRadius)
        path.rotate(to: 36.0, around: starCenter)
        path.fill()
        // Funafuti island
        starCenter = CGPoint(x: size.width / 1000.0 * 841.5, y: size.height / 100.0 * 29.3)
        path = UIBezierPath.star(at: starCenter, radius: starRadius)
        path.rotate(to: 36.0, around: starCenter)
        path.fill()

        flagUnitedKingdom(with: size.halfSize).draw(at: .zero)

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
